sway-core/src/type_system/type_mapping.rs

```diff
@@ -350,10 +357,13 @@ impl TypeMapping {
                     None
                 }
             }
-            TypeInfo::Array(ary_ty_id, count, initial_elem_ty) => {
-                self.find_match(ary_ty_id, type_engine).map(|matching_id| {
-                    type_engine.insert_type(TypeInfo::Array(matching_id, count, initial_elem_ty))
-                })
+            TypeInfo::Array(mut elem_ty, count) => {
+                if let Some(type_id) = self.find_match(elem_ty.type_id, type_engine) {
+                    elem_ty.type_id = type_id;
+                    Some(type_engine.insert_type(TypeInfo::Array(elem_ty, count)))
+                } else {
+                    None
+                }
```

Comment:
```suggestion
            TypeInfo::Array(mut elem_ty, count) => self
                .find_match(elem_ty.type_id, type_engine)
                .map(|type_id| {
                    elem_ty.type_id = type_id;
                    type_engine.insert_type(TypeInfo::Array(elem_ty, count))
                }),
```
This could've stayed as a `map`.